src/main/java/io/cdap/plugin/http/source/batch/HttpBatchSource.java

```diff
@@ -0,0 +1,146 @@
+import io.cdap.plugin.http.source.common.HttpErrorHandler;
+import io.cdap.plugin.http.source.common.HttpErrorHandlingStrategy;
+import io.cdap.plugin.http.source.common.record.BaseStringToRecordConverter;
+import io.cdap.plugin.http.source.common.record.StringToRecordConverterFactory;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.stream.Collectors;
+
+/**
+ * Plugin returns records from HTTP source specified by link. Pagination via APIs is supported.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(HttpBatchSource.NAME)
+@Description("Read data from HTTP endpoint.")
+public class HttpBatchSource extends BatchSource<LongWritable, Text, StructuredRecord> {
+  static final String NAME = "HTTP";
+  private static final Logger LOG = LoggerFactory.getLogger(HttpBatchSource.class);
+
+  private static final String ERROR_SCHEMA_BODY_PROPERTY = "body";
+
+  private static final Schema errorSchema = Schema.recordOf("error",
+    Schema.Field.of(ERROR_SCHEMA_BODY_PROPERTY, Schema.of(Schema.Type.STRING))
+  );
+
+  private final HttpBatchSourceConfig config;
+  private Schema schema;
+  private HttpErrorHandler httpErrorHandler;
+
+  public HttpBatchSource(HttpBatchSourceConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    config.validate(); // validate when macros not yet substituted
+    config.validateSchema();
+
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(config.getSchema());
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) {
+    config.validate(); // validate when macros are already substituted
+    config.validateSchema();
+
+    if (schema == null) {
+      schema = config.getSchema();
+    }
+
+    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
+    lineageRecorder.createExternalDataset(schema);
+    lineageRecorder.recordRead("Read", String.format("Read from HTTP '%s'", config.getUrl()),
+      Preconditions.checkNotNull(schema.getFields()).stream()
+        .map(Schema.Field::getName)
+        .collect(Collectors.toList()));
+
+    context.setInput(Input.of(config.referenceName, new HttpInputFormatProvider(config)));
+  }
+
+  private BaseStringToRecordConverter stringToRecordConverter;
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    this.schema = context.getOutputSchema();
+    this.stringToRecordConverter = StringToRecordConverterFactory.createInstance(config, schema);
+    this.httpErrorHandler = new HttpErrorHandler(config);
+
+    super.initialize(context);
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Text> input,
+                        Emitter<StructuredRecord> emitter) throws IOException {
+    String resultString = input.getValue().toString();
```

Comment:
I just tried to do this, but it makes things a bit to over-complicated when error handling strategy is "SKIP". Since we cannot simply skip in iterator, if has hasNext() already returned. We can pre-cache of course. I don't really like that.

I would prefer to have here "WritableComparable" instead of "Text" when we implement Blob.
Both BytesWritable and Text implement WritableComparable.

What do you think?

Reply:
I'm not sure I understand why it's hard to skip in the RecordReader, but sending to error seems like it would be impossible. 

Not sure what would be the best way right now, let's revisit when formats are implemented.